# 1 - Introdução

This Online Retail II data set contains all the transactions occurring for a UK-based and registered, non-store online retail between 01/12/2009 and 09/12/2011.The company mainly sells unique all-occasion gift-ware. Many customers of the company are wholesalers.

## 1.1 - Dicionário de dados

**InvoiceNo:** Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation. 

**StockCode:** Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product. 

**Description:** Product (item) name. Nominal. 

**Quantity:** The quantities of each product (item) per transaction. Numeric.

**InvoiceDate:** Invice date and time. Numeric. The day and time when a transaction was generated. 

**UnitPrice:** Unit price. Numeric. Product price per unit in sterling (£). 

**CustomerID:** Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer. 

**Country:** Country name. Nominal. The name of the country where a customer resides.

In [2]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [3]:
%%time

# O arquivo veio em formato xlsx (planilha excel). Nele, há 2 abas, cada uma representando um ano.
# Obs.: para carregar o conjunto de dados, a planilha não pode estar aberta.

# Ano 2009-2010
df_0 = pd.read_excel('online_retail_II.xlsx', sheet_name=0)

# Ano 2010-2011
df_1 = pd.read_excel('online_retail_II.xlsx', sheet_name=1)

CPU times: total: 6min 21s
Wall time: 7min 3s


In [4]:
df_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [5]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [6]:
# Os dataframes apresentam a mesma estrutura de dados (colunas iguais), portanto eles serão mesclados.
df = pd.concat([df_0, df_1], ignore_index = True)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[ns]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 65.1+ MB


In [7]:
df.sample(5)

Invoice StockCode                        Description  Quantity  \
51318   494019     84836        ZINC METAL HEART DECORATION         1   
510666  537051     21916    SET 12 RETRO WHITE CHALK STICKS         4   
734652  555194     21988       PACK OF 6 SKULL PAPER PLATES        96   
884595  568176     22720  SET OF 3 CAKE TINS PANTRY DESIGN          3   
468675  533554     21162           TOXIC AREA  DOOR HANGER          4   

               InvoiceDate  Price  Customer ID         Country  
51318  2010-01-11 09:38:00   2.57          NaN  United Kingdom  
510666 2010-12-05 11:12:00   0.42      15708.0  United Kingdom  
734652 2011-06-01 11:13:00   0.64      16422.0  United Kingdom  
884595 2011-09-25 13:25:00   4.95      12685.0          France  
468675 2010-11-17 17:16:00   1.45      13605.0  United Kingdom

In [8]:
# Salvando o resultado da mescla das planilhas em um arquivo csv.
df.to_csv('dataset_backup.csv', index = False)

# 2 - Manipulação e tratamento dos dados

## 2.1 - Tipo dos dados

Inicialmente, será feita uma análise com relação aos tipos de dados de cada coluna do dataset.

In [51]:
# Recarregando o conjunto de dados.
df = pd.read_csv('dataset_backup.csv')

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  object 
 1   StockCode    1067371 non-null  object 
 2   Description  1062989 non-null  object 
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  object 
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 65.1+ MB


> A coluna `Customer ID` apresenta dados do tipo inteiro. Porém isso não condiz com a informação contida nela. Além disso, após carregar o conjunto de dados anteriormente salvo, a coluna de data (`InvoiceDate`) perdeu a formatação original (datetime). Portanto, nessa etapa, realizaremos duas mudanças:

1. `InvoiceDate`: de object para datetime
2. `Customer ID`: de inteiro para object

In [53]:
# Mudando o tipo da coluna InvoiceDate.
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], errors='coerce')

In [54]:
# Mudando o tipo da coluna Customer ID.
# Utilizou-se o método map, com o parâmetro na_action = ignore, para manter os valores nulos dessa coluna.

df['Customer ID'] = df['Customer ID'].map(str, na_action = 'ignore')

In [13]:
df.sample(5)

Invoice StockCode                         Description  Quantity  \
708642  552652     22700            BLACK AND WHITE DOG BOWL         1   
754935  557057     22385           JUMBO BAG SPACEBOY DESIGN        10   
854831  565842     21813        GARLAND WITH STARS AND BELLS         6   
327356  521357     21791  VINTAGE HEADS AND TAILS CARD GAME          1   
187996  507235    72801D         S/4 SKY BLUE DINNER CANDLES        24   

               InvoiceDate  Price Customer ID         Country  
708642 2011-05-10 14:09:00   2.95     14591.0  United Kingdom  
754935 2011-06-16 14:46:00   2.08     16843.0  United Kingdom  
854831 2011-09-07 12:10:00   4.95     13808.0  United Kingdom  
327356 2010-09-03 16:04:00   2.51         NaN  United Kingdom  
187996 2010-05-07 09:04:00   1.06     14298.0  United Kingdom

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[ns]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   object        
 7   Country      1067371 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 65.1+ MB


In [55]:
# Separando as colunas com dados numéricos e dados do tipo string 

# Colunas numéricas
numericos = ['Quantity', 'Price']

# Colunas "categóricas"
cat = ['Invoice', 'StockCode', 'Description', 'Customer ID', 'Country']

## 2.2 - Manipulação dos dados

Analisaremos as colunas `Quantity` e `Price`. Para isso, utilizaremos o método describe(), que nos retorna algumas estatísticas descritivas dos dados.

In [16]:
# Análise das colunas numéricas
round(df[numericos].describe(), 2)

Quantity       Price
count  1067371.00  1067371.00
mean         9.94        4.65
std        172.71      123.55
min     -80995.00   -53594.36
25%          1.00        1.25
50%          3.00        2.10
75%         10.00        4.15
max      80995.00    38970.00

> Os seguintes valores da tabela acima serão analisados:
   * Quantidade negativa
   * Preço negativo
   * Preço igual à zero
    
> Além disso, as faturas com cancelamento (que possuem a letra "C" na coluna `Invoice`) também serão analisadas.

### 2.2.1 - Quantidades negativas

In [56]:
df_qtd_negativo = df[df['Quantity'] < 0]

In [57]:
df_qtd_negativo.sample(10)

Invoice StockCode                      Description  Quantity  \
371772   525266     45016                              NaN       -36   
940781  C572538     22456  NATURAL SLATE CHALKBOARD LARGE         -3   
904094  C569655     22729      ALARM CLOCK BAKELIKE ORANGE        -2   
240230  C512592     21623     VINTAGE UNION JACK MEMOBOARD        -2   
355942  C523974     21844                RED RETROSPOT MUG        -1   
215022  C510214     21843         RED RETROSPOT CAKE STAND       -60   
708182  C552564     23201               JUMBO BAG ALPHABET      -200   
904089  C569655     21395            BLUE POLKADOT BEAKER         -6   
5426    C489858     22198           POPCORN HOLDER , LARGE        -7   
675700   549373     85008                              NaN        -2   

               InvoiceDate  Price Customer ID         Country  
371772 2010-10-04 16:13:00   0.00         NaN  United Kingdom  
940781 2011-10-24 16:10:00   4.95     14121.0  United Kingdom  
904094 2011-10-05 12:58:00   3.75     14534.0  United Kingdom  
240230 2010-06-16 16:34:00   9.95     16347.0  United Kingdom  
355942 2010-09-26 13:59:00   2.95     15513.0  United Kingdom  
215022 2010-05-28 08:16:00   9.95     15061.0  United Kingdom  
708182 2011-05-10 11:55:00   1.79     13694.0  United Kingdom  
904089 2011-10-05 12:58:00   0.39     14534.0  United Kingdom  
5426   2009-12-02 14:43:00   1.65     12471.0         Germany  
675700 2011-04-08 11:28:00   0.00         NaN  United Kingdom

> Após uma análise mais aprofundada das descrições das fatura, foi possível descobrir:
1. Existem faturas que não são cancelamentos, mas apresentam a descrição de "damage" ou "missing". Porém, aparentemente, essa coluna deve ser preenchida de forma manual, pois existem escritas diferentes para a mesma informação (damage e Damage, por exemplo)
2. A maior parte das faturas com quantidades negativas parecem ser de cancelamentos (isso será analisado posteriormente)

In [58]:
# Quantidade de registros com valores negativos:
df_qtd_negativo.shape

(22950, 8)

#### a) Faturas com descrição "damage"

In [59]:
damage = ['Damage', 'damage'] # abordando formas diferentes de escrita

# para o df (todos os registros)
df[df['Description'].str.contains('|'.join(damage), na=False)].shape

(218, 8)

In [35]:
df[df['Description'].str.contains('|'.join(damage), na=False)].sample(5)

Invoice StockCode    Description  Quantity         InvoiceDate  Price  \
194007  507873     20820        damages      -143 2010-05-11 14:59:00    0.0   
736421  555331    84800L    wet damaged      -144 2011-06-02 11:11:00    0.0   
991158  576250     71279        damages       -74 2011-11-14 13:50:00    0.0   
224966  511116     85060  damaged/dirty      -408 2010-06-07 10:40:00    0.0   
736424  555334    84805A    wet damaged       -96 2011-06-02 11:12:00    0.0   

       Customer ID         Country  
194007         NaN  United Kingdom  
736421         NaN  United Kingdom  
991158         NaN  United Kingdom  
224966         NaN  United Kingdom  
736424         NaN  United Kingdom

> Portanto, encontramos um padrão nos dados. Faturas que apresentam avarias possuem preço zerado e não possuem clientes (`Customer ID`). Criaremos uma nova coluna `status` e as faturas que se encaixam nesse padrão terão como valor "Damage".

In [60]:
# Criando uma coluna vazia (coma valores NaN. Essa coluna será preechida ao longo das manipulações)
df['Status'] = np.nan

In [61]:
# Adicionando a categoria "Damage" para a nova coluna.
df['Status'] = np.where(df['Description'].str.contains('|'.join(damage), na=False), "Damage", np.nan)

In [38]:
df.sample(5)

Invoice StockCode                       Description  Quantity  \
555724  538853     20854       BLUE PATCH PURSE PINK HEART         2   
52698   494213     21328            BALLOONS  WRITING SET          1   
604218  542898     22624              IVORY KITCHEN SCALES         1   
874103  567461    84997b  CHILDRENS CUTLERY RETROSPOT RED          3   
772409  558749     23284     DOORMAT KEEP CALM AND COME IN         3   

               InvoiceDate  Price Customer ID         Country Status  
555724 2010-12-14 13:35:00   1.65     16805.0  United Kingdom    nan  
52698  2010-01-12 11:40:00   1.65     17945.0  United Kingdom    nan  
604218 2011-02-01 14:01:00   8.50     16110.0  United Kingdom    nan  
874103 2011-09-20 12:31:00   8.29         NaN  United Kingdom    nan  
772409 2011-07-03 10:28:00   7.95     16326.0  United Kingdom    nan

#### b) Faturas com descrição "missing"

In [62]:
missing = ['missing', 'Missing'] # abordando formas diferentes de escrita

# para o df (todos os registros)
df[df['Description'].str.contains('|'.join(missing), na=False)].shape

(39, 9)

In [63]:
df[df['Description'].str.contains('|'.join(missing), na=False)].sample(5)

Invoice StockCode               Description  Quantity  \
47635    493821    17129D  missing (wrongly coded?)     -2127   
990303   576185     22575                  missing?      -130   
1023280  578476    72807B               ????missing      -124   
122587   501109     77081                   missing      -233   
1019147  578245     22568                ?? missing      -170   

                InvoiceDate  Price Customer ID         Country Status  
47635   2010-01-07 12:43:00    0.0         NaN  United Kingdom    nan  
990303  2011-11-14 11:25:00    0.0         NaN  United Kingdom    nan  
1023280 2011-11-24 12:45:00    0.0         NaN  United Kingdom    nan  
122587  2010-03-12 13:44:00    0.0         NaN  United Kingdom    nan  
1019147 2011-11-23 12:37:00    0.0         NaN  United Kingdom    nan

In [64]:
# Adicionando a categoria "Missing" para a nova coluna.
df['Status'] = np.where(df['Description'].str.contains('|'.join(damage), na=False), "Damage",
                        np.where(df['Description'].str.contains('|'.join(missing), na=False), "Missing",
                        np.nan))

#### c) Cancelamentos

In [65]:
# Inicialmente, analisaremaos todas as faturas com cancelamento
df_cancel = df[df['Invoice'].str.contains('C', na=False)]
df_cancel.shape

(19494, 9)

In [66]:
df['Status'] = np.where(df['Description'].str.contains('|'.join(damage), na=False), "Damage",
                        np.where(df['Description'].str.contains('|'.join(missing), na=False), "Missing",
                        np.where(df['Invoice'].str.contains('C', na=False), "Cancellation",
                        np.nan)))

In [67]:
# Recarregando o dataset com quantidades negativas, agora com a coluna Status.
df_qtd_negativo = df[df['Quantity'] < 0]

In [68]:
df_qtd_negativo_damage = df_qtd_negativo[df_qtd_negativo['Status'] == 'Damage']
df_qtd_negativo_missing = df_qtd_negativo[df_qtd_negativo['Status'] == 'Missing']
df_qtd_negativo_cancel = df_qtd_negativo[df_qtd_negativo['Status'] == 'Cancellation']

print('Para quantidades negativas:\n')
print(f'Faturas com avarias: {df_qtd_negativo_damage.shape[0]}')
print(f'Faturas perdidas: {df_qtd_negativo_missing.shape[0]}')
print(f'Faturas canceladas: {df_qtd_negativo_cancel.shape[0]}')

Para quantidades negativas:

Faturas com avarias: 215
Faturas perdidas: 39
Faturas canceladas: 19493


> Em resumo:
    Temos um total de **22.950** registros que apresentam quantidades negativas. Desses registros:
   * **215** são referentes à avarias e foram categorizados como "Damage" na coluna `Status`;
   * **39** são de produtos que se perderam (missing). Foram categorizados como "Missing" na coluna `Status`
   * **19.493** são de cancelamentos.
   * Ainda temos **3.203** registros não categorizados

In [69]:
# -----------------------------------------------------------------------------------------------------------------------------
# 3. Registros não categorizados (registros com quantidades negativas mas que não são cancelamentos nem possuem avarias)

df_qtd_negativo = df[df['Quantity'] < 0]
df_nao_cat = df_qtd_negativo[(df_qtd_negativo['Status'] != 'Damage') &
                             (df_qtd_negativo['Status'] != 'Cancellation') &
                            (df_qtd_negativo['Status'] != 'Missing')]

In [70]:
df_nao_cat.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3203 entries, 263 to 1064386
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Invoice      3203 non-null   object        
 1   StockCode    3203 non-null   object        
 2   Description  514 non-null    object        
 3   Quantity     3203 non-null   int64         
 4   InvoiceDate  3203 non-null   datetime64[ns]
 5   Price        3203 non-null   float64       
 6   Customer ID  0 non-null      object        
 7   Country      3203 non-null   object        
 8   Status       3203 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 250.2+ KB


> Para esses registros (**3.203**), como demonstrado acima, todos eles não possuem Id de clientes e a grande maioria não apresenta descrição.

In [71]:
# Iremos observar algumas descrições para esse conjunto de dados

df_nao_cat.drop("Customer ID", inplace = True, axis = 1)

In [72]:
df_nao_cat.dropna(inplace = True, axis = 0)
df_nao_cat.sample(10)

Invoice StockCode             Description  Quantity  \
136410  502458     84016              ebay sales     -5000   
916179  570595     21823                   check       -94   
848923  565307     21830                       ?      -443   
804712  561249  DCGS0073                    ebay        -4   
665661  548381     22127            sold in set?      -690   
154171  503992    35821B                    gone       -21   
954080  573490    85035A                   CHECK       -59   
900888  569464     23268  Incorrect stock entry.     -1440   
437008  531177     20852              Given away     -8800   
265653  515049     35963                 checked       -10   

               InvoiceDate  Price         Country Status  
136410 2010-03-24 14:21:00    0.0  United Kingdom    nan  
916179 2011-10-11 11:47:00    0.0  United Kingdom    nan  
848923 2011-09-02 12:20:00    0.0  United Kingdom    nan  
804712 2011-07-26 11:51:00    0.0  United Kingdom    nan  
665661 2011-03-30 16:45:00    0.0  United Kingdom    nan  
154171 2010-04-09 11:02:00    0.0  United Kingdom    nan  
954080 2011-10-31 11:55:00    0.0  United Kingdom    nan  
900888 2011-10-04 11:40:00    0.0  United Kingdom    nan  
437008 2010-11-05 14:40:00    0.0  United Kingdom    nan  
265653 2010-07-08 10:50:00    0.0  United Kingdom    nan

> Por fim, é possível observar que ainda existem alguns padrões nos dados (por exemplo, a descrição "Check"). Observando as descrições que existiam (**514** de **3.203**), podemos perceber que são faturas que provavelmente foram preenchidas de forma manual, sem a devida padronização. Por ser uma quantidade baixa de registros, em comparação ao total, eles serão excluídos do conjunto de dados, sem perdas para as análises.

In [73]:
# Recarregando o dataset df_nao_cat, pois os valores nulos haviam sido suprimidos
df_nao_cat = df_qtd_negativo[(df_qtd_negativo['Status'] != 'Damage') &
                             (df_qtd_negativo['Status'] != 'Cancellation') &
                            (df_qtd_negativo['Status'] != 'Missing')]

# Realizando o drop, por índice das linhas do dataset acima, ou seja, removendo o dataset df_nao_cat do original
df.drop(df_nao_cat.index.values, axis = 0, inplace = True)

### 2.2.2 - Preços negativos

In [74]:
df_prc_negativo = df[df['Price'] < 0]

In [75]:
df_prc_negativo.sample(5)

Invoice StockCode      Description  Quantity         InvoiceDate  \
179403  A506401         B  Adjust bad debt         1 2010-04-29 13:36:00   
403472  A528059         B  Adjust bad debt         1 2010-10-20 12:04:00   
276274  A516228         B  Adjust bad debt         1 2010-07-19 11:24:00   
825444  A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
825445  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           Price Customer ID         Country Status  
179403 -53594.36         NaN  United Kingdom    nan  
403472 -38925.87         NaN  United Kingdom    nan  
276274 -44031.79         NaN  United Kingdom    nan  
825444 -11062.06         NaN  United Kingdom    nan  
825445 -11062.06         NaN  United Kingdom    nan

In [76]:
df_prc_negativo.shape

(5, 9)

> Temos apenas 5 registros com preços negativos. Esses registros apresentam em sua descrição "Adjust bad debt" ou "Ajustar dívidas inadimplentes". Portanto, na coluna `Status`, eles serão preenchidos com a informação "Ajust bad debt"

In [77]:
df['Status'] = np.where(df['Description'].str.contains('|'.join(damage), na=False), "Damage",
                        np.where(df['Description'].str.contains('|'.join(missing), na=False), "Missing",
                        np.where(df['Invoice'].str.contains('C', na=False), "Cancellation",
                        np.where((df['Price'] < 0), "Ajust bad debt",         
                        np.nan))))

### 2.2.3 - Preços iguais à zero (excluindo registros com avarias e perdidos)

In [78]:
df_prc_zero = df[(df['Price'] == 0) & (df['Status'] != "Damage") & (df['Status'] != "Missing")]
df_prc_zero.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2742 entries, 3161 to 1064015
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Invoice      2742 non-null   object        
 1   StockCode    2742 non-null   object        
 2   Description  1049 non-null   object        
 3   Quantity     2742 non-null   int64         
 4   InvoiceDate  2742 non-null   datetime64[ns]
 5   Price        2742 non-null   float64       
 6   Customer ID  71 non-null     object        
 7   Country      2742 non-null   object        
 8   Status       2742 non-null   object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 214.2+ KB


In [79]:
df_prc_zero['Status'].value_counts()

nan    2742
Name: Status, dtype: int64

In [80]:
df_prc_zero

Invoice StockCode                   Description  Quantity  \
3161     489659     21350                           NaN       230   
3731     489781     84292                           NaN        17   
4674     489825     22076            6 RIBBONS EMPIRE          12   
5904     489861       DOT                DOTCOM POSTAGE         1   
6378     489882    35751C                           NaN        12   
...         ...       ...                           ...       ...   
1060795  581211     22142                         check        14   
1062442  581234     72817                           NaN        27   
1063965  581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
1063966  581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
1064015  581408     85175                           NaN        20   

                InvoiceDate  Price Customer ID         Country Status  
3161    2009-12-01 17:39:00    0.0         NaN  United Kingdom    nan  
3731    2009-12-02 11:45:00    0.0         NaN  United Kingdom    nan  
4674    2009-12-02 13:34:00    0.0     16126.0  United Kingdom    nan  
5904    2009-12-02 14:50:00    0.0         NaN  United Kingdom    nan  
6378    2009-12-02 16:22:00    0.0         NaN  United Kingdom    nan  
...                     ...    ...         ...             ...    ...  
1060795 2011-12-07 18:36:00    0.0         NaN  United Kingdom    nan  
1062442 2011-12-08 10:33:00    0.0         NaN  United Kingdom    nan  
1063965 2011-12-08 13:58:00    0.0         NaN  United Kingdom    nan  
1063966 2011-12-08 13:58:00    0.0         NaN  United Kingdom    nan  
1064015 2011-12-08 14:06:00    0.0         NaN  United Kingdom    nan  

[2742 rows x 9 columns]

> temos alguns registros que apresentam preço igual à zero e que não se enquadram em nenhuma categoria da coluna `Status`. Percebemos valores nulos, para a grande maioria, nas colunas `Customer ID` e `Description`. Esses registros também serão excluídos do conjunto de dados, por não possuírem um tipo de padrão. 

In [81]:
df.drop(df_prc_zero.index.values, axis = 0, inplace = True)

> Por fim, os registros que não foram considerados danificados, perdidos, ajuste de dívida ou cancelamento, são faturas normais, resultante de vendas e terão o status "Ok".

In [82]:
df['Status'] = np.where(df['Description'].str.contains('|'.join(damage), na=False), "Damage",
                        np.where(df['Description'].str.contains('|'.join(missing), na=False), "Missing",
                        np.where(df['Invoice'].str.contains('C', na=False), "Cancellation",
                        np.where((df['Price'] < 0), "Ajust bad debt",         
                        'Ok'))))

In [83]:
df['Status'].value_counts()

Ok                1041670
Cancellation        19494
Damage                218
Missing                39
Ajust bad debt          5
Name: Status, dtype: int64

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1061426 entries, 0 to 1067370
Data columns (total 9 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1061426 non-null  object        
 1   StockCode    1061426 non-null  object        
 2   Description  1061426 non-null  object        
 3   Quantity     1061426 non-null  int64         
 4   InvoiceDate  1061426 non-null  datetime64[ns]
 5   Price        1061426 non-null  float64       
 6   Customer ID  824293 non-null   object        
 7   Country      1061426 non-null  object        
 8   Status       1061426 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 81.0+ MB


### 2.3 - Tratamento de valores missing

Somente a coluna `Customer ID` apresenta valores missing. Temos uma quantidade grande de dados nulos (mais de 22%), o que representaria uma grande perda para a análise e também a supressão de informações dos KPIs que analisaremos mais à frente. Portanto, seguiremos com esses valores nulos para as próximas etapas. Porém, ao analisarmos o conjunto de dados à nível de clientes, esses registros precisarão ser suprimidos.

In [87]:
print('Porcentagem de valores missing, por coluna:\n')
print(round((df.isnull().sum()/df.shape[0])*100, 1))

Porcentagem de valores missing, por coluna:

Invoice         0.0
StockCode       0.0
Description     0.0
Quantity        0.0
InvoiceDate     0.0
Price           0.0
Customer ID    22.3
Country         0.0
Status          0.0
dtype: float64


In [67]:
df[df['Customer ID'].isnull()]

Invoice StockCode                     Description  Quantity  \
577      489525    85226C       BLUE PULL BACK RACING CAR         1   
578      489525     85227     SET/6 3D KIT CARDS FOR KIDS         1   
1055     489548     22271            FELTCRAFT DOLL ROSIE         1   
1056     489548     22254            FELT TOADSTOOL LARGE        12   
1057     489548     22273            FELTCRAFT DOLL MOLLY         3   
...         ...       ...                             ...       ...   
1066997  581498    85099B         JUMBO BAG RED RETROSPOT         5   
1066998  581498    85099C  JUMBO  BAG BAROQUE BLACK WHITE         4   
1066999  581498     85150   LADIES & GENTLEMEN METAL SIGN         1   
1067000  581498     85174               S/4 CACTI CANDLES         1   
1067001  581498       DOT                  DOTCOM POSTAGE         1   

                InvoiceDate    Price Customer ID         Country Status  
577     2009-12-01 11:49:00     0.55         NaN  United Kingdom     Ok  
578     2009-12-01 11:49:00     0.85         NaN  United Kingdom     Ok  
1055    2009-12-01 12:32:00     2.95         NaN  United Kingdom     Ok  
1056    2009-12-01 12:32:00     1.25         NaN  United Kingdom     Ok  
1057    2009-12-01 12:32:00     2.95         NaN  United Kingdom     Ok  
...                     ...      ...         ...             ...    ...  
1066997 2011-12-09 10:26:00     4.13         NaN  United Kingdom     Ok  
1066998 2011-12-09 10:26:00     4.13         NaN  United Kingdom     Ok  
1066999 2011-12-09 10:26:00     4.96         NaN  United Kingdom     Ok  
1067000 2011-12-09 10:26:00    10.79         NaN  United Kingdom     Ok  
1067001 2011-12-09 10:26:00  1714.17         NaN  United Kingdom     Ok  

[236979 rows x 9 columns]

### 2.4 - Conclusão

Resumindo, até o momento foram realizadas algumas manipulações e tratamentos nos dados, com o intuito de se ter um maior entendimento deles. Com isso, conseguimos agrupar os registros em 5 diferentes status:
1. Ok: registros resultado de vendas de sucesso. Serão contabilizado no faturamento da empresa.
2. Cancellation: registros decorrentes de cancelamentos. Eles serão descontados do faturamento da empresa, por possuir quantidade negativa*
3. Damage: resultado de produtos com avarias. Não influenciarão no faturamento por possuir preço = 0.
4. Missing: resultado de produtos perdidos. Não influenciarão no faturamento por possuir preço = 0.
5. Ajust bad debt: ajustes de dídiva de inadimplência. Serão descontados do faturamento.

In [91]:
df[df["Status"] == 'Missing'][numericos].describe()

Quantity  Price
count    39.000000   39.0
mean   -592.435897    0.0
std     740.041777    0.0
min   -3340.000000    0.0
25%    -598.000000    0.0
50%    -256.000000    0.0
75%    -170.000000    0.0
max     -65.000000    0.0

In [92]:
df[df["Status"] == 'Damage'][numericos].describe()

Quantity  Price
count   218.000000  218.0
mean   -156.307339    0.0
std     418.554680    0.0
min   -4354.000000    0.0
25%     -96.000000    0.0
50%     -41.000000    0.0
75%     -15.000000    0.0
max     192.000000    0.0

In [95]:
df[(df["Status"] == 'Cancellation') & (df["Quantity"] > 0)]

Invoice StockCode Description  Quantity         InvoiceDate   Price  \
76799  C496350         M      Manual         1 2010-02-01 08:24:00  373.57   

      Customer ID         Country        Status  
76799         NaN  United Kingdom  Cancellation